consensus/hotstuff/verification/combined_verifier_v2.go

```diff
@@ -0,0 +1,142 @@
+// +build relic
+
+package verification
+
+import (
+	"fmt"
+
+	"github.com/onflow/flow-go/consensus/hotstuff"
+	"github.com/onflow/flow-go/consensus/hotstuff/model"
+	"github.com/onflow/flow-go/consensus/hotstuff/signature"
+	"github.com/onflow/flow-go/crypto"
+	"github.com/onflow/flow-go/crypto/hash"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module"
+	modulesig "github.com/onflow/flow-go/module/signature"
+)
+
+// CombinedVerifier is a verifier capable of verifying two signatures for each
+// verifying operation. The first type is a signature from an aggregating signer,
+// which verifies either the single or the aggregated signature. The second type is
+// a signature from a threshold signer, which verifies either the signature share or
+// the reconstructed threshold signature.
+type CombinedVerifierV2 struct {
+	committee      hotstuff.Committee
+	staking        hash.Hasher
+	beacon         hash.Hasher
+	keysAggregator *stakingKeysAggregator
+	merger         module.Merger
+}
+
+// NewCombinedVerifier creates a new combined verifier with the given dependencies.
+// - the hotstuff committee's state is used to retrieve the public keys for the staking signature;
+// - the staking tag is used to create hasher to verify staking signatures;
+// - the beacon tag is used to create hasher to verify random beacon signatures;
+// - the merger is used to combined & split staking & random beacon signatures;
+func NewCombinedVerifierV2(committee hotstuff.Committee, stakingTag string, beaconTag string, merger module.Merger) *CombinedVerifierV2 {
+	return &CombinedVerifierV2{
+		committee:      committee,
+		staking:        crypto.NewBLSKMAC(stakingTag),
+		beacon:         crypto.NewBLSKMAC(beaconTag),
+		keysAggregator: newStakingKeysAggregator(),
+		merger:         merger,
+	}
+}
+
+// VerifyVote verifies the validity of a combined signature from a vote.
+// Usually this method is only used to verify the proposer's vote, which is
+// the vote included in a block proposal.
+// TODO: return error only, because when the sig is invalid, the returned bool
+// can't indicate whether it's staking sig was invalid, or beacon sig was invalid.
+func (c *CombinedVerifierV2) VerifyVote(signer *flow.Identity, sigData []byte, block *model.Block) (bool, error) {
+
+	// create the to-be-signed message
+	msg := MakeVoteMessage(block.View, block.BlockID)
+
+	// split the two signatures from the vote
+	// TODO: move DecodeDoubleSig to merger.Split
+	stakingSig, beaconShare, err := signature.DecodeDoubleSig(sigData)
+	if err != nil {
+		return false, fmt.Errorf("could not split signature: %w", modulesig.ErrInvalidFormat)
+	}
+
+	dkg, err := c.committee.DKG(block.BlockID)
+	if err != nil {
+		return false, fmt.Errorf("could not get dkg: %w", err)
+	}
+
+	// get the signer dkg key share
+	beaconPubKey, err := dkg.KeyShare(signer.NodeID)
```

Comment:
if we have failed DKG and we don't have a private key, will we have a public key in such situation?
We can move this down the path, after checking if we have `beaconShare` this way we will localize beacon sig checks and staking sig checks

Reply:
Good catch @durkmurder 

Also a question for @tarakby:

Say there are `A,B,C,D,E,F,G` nodes, nodeA completed the DKG, but only node A failed to generate the DKG private key, all other nodes successfully generated their DKG private key. Then does node C who succeeded, know that A failed? or node C will have a special public key for nodeA but no one knows its private key?

The reason I'm asking is that, if node C knows nodeA has succeeded DKG or not, then it can reject votes proposals or votes from node A that doesn't have random beacon key share.

Reply:
> if we have failed DKG and we don't have a private key, will we have a public key in such situation? 

If DKG failed globally then there will be no keys at all, but the epoch transition would have failed earlier before we get to Hotstuff of the new epoch (code above). If we get to this code, then there will be always public keys, regardless of the local private key result. 

> Then does node C who succeeded, know that A failed DKG?
C would have its local subjective view of how DKG ended for everyone (A and others). Though its local view may not correspond to what the objective reality is. That's why we do not rely on it. 
There is one case where all nodes can be sure that A failed, though that is not implemented yet (it can be useful to make everyone skip A's beacon signatures when they know A failed)